mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -0,0 +1,1758 @@
+  // scf.for %arg0 = %c0 to %c128 step %c1 {
+  //  scf.for %arg1 = %c0 to %c30 step %c1 {
+  //    scf.for %arg2 = %c0 to %c30 step %c1 {
+  //      scf.for %arg3 = %c0 to %c1 step %c1 {
+  //        scf.for %arg4 = %c0 to %c128 step %c1 {
+  //          %2 = load %CpuResults[%arg0, %arg1, %arg2, %arg3, %arg4] : memref<128x30x30x1x128xf32>
+  //          %3 = load %GpuResults[%arg0, %arg1, %arg2, %arg3, %arg4] : memref<128x30x30x1x128xf32>
+  //          %cst = constant 1.000000e-07 : f32
+  //          %4 = subf %2, %3 : f32
+  //          %5 = absf %4 : f32
+  //          %6 = cmpf ugt, %5, %cst : f32
+  //          scf.if %6 {
+  //            store %c0_i32, %result[%c0] : memref<1xi32>
+  //          }
+  //        }
+  //      }
+  //    }
+  //  }
+  //}
+  // clang-format on
+
+  SmallVector<mlir::Value, 5> dimOps;
+  for (uint32_t i = 0; i < dims.size(); ++i) {
+    dimOps.push_back(b.create<arith::ConstantIndexOp>(loc, dims[i]));
+  }
+
+  scf::ForOp loop;
+  OpBuilder loopB = b;
+  SmallVector<mlir::Value, 5> idxs;
+  for (uint32_t i = 0; i < dims.size(); ++i) {
+    loop = loopB.create<scf::ForOp>(loc, c0IndexOp, dimOps[i], c1IndexOp);
+    loopB = OpBuilder::atBlockTerminator(loop.getBody());
+    idxs.push_back(loop.getInductionVar());
+  }
+
+  // Inner loop body
+  mlir::Value gpuLoadVal =
+      loopB.create<memref::LoadOp>(loc, block->getArgument(0), idxs);
+  mlir::Value cpuLoadVal =
+      loopB.create<memref::LoadOp>(loc, block->getArgument(1), idxs);
+
+  mlir::Value gpuFPVal = gpuLoadVal;
+  mlir::Value cpuFPVal = cpuLoadVal;
+
+  // Lower cpu values to gpu precision
+  mlir::FloatType elemFType = floatType;
+  if (!elemType.isF32()) {
+    if (elemType.isInteger(16)) {
+      gpuLoadVal =
+          loopB.create<arith::BitcastOp>(loc, gpuLoadVal, loopB.getBF16Type());
+      elemFType = b.getBF16Type();
+    } else {
+      elemFType = b.getF16Type();
+    }
+    cpuLoadVal = loopB.create<arith::TruncFOp>(loc, cpuLoadVal, elemFType);
+    gpuFPVal = loopB.create<arith::ExtFOp>(loc, gpuLoadVal, floatType);
+  }
+
+  mlir::Value percentDiffVal;
+  mlir::Value cmpVal;
+  if ((randomSeed.getValue() != "none" &&
+       randomDataType.getValue() == "float") ||
+      !elemType.isF32()) {
+    // <test> = |(<cpu> - <gpu>) / <cpu>| > <max_percent>
+
+    auto getFVal = [&](float val) -> mlir::Value {
+      llvm::APFloat apVal(val);
+      bool ignored;
+      if (elemFType.isF16()) {
+        apVal.convert(llvm::APFloat::IEEEhalf(), APFloat::rmNearestTiesToEven,
+                      &ignored);
+      } else if (elemFType.isBF16()) {
+        apVal.convert(llvm::APFloat::BFloat(), APFloat::rmNearestTiesToEven,
+                      &ignored);
+      }
+      return loopB.create<arith::ConstantFloatOp>(loc, apVal, elemFType);
+    };
+
+    // <test> = <cpu> - <gpu>
+    auto subfOp = loopB.create<arith::SubFOp>(loc, cpuLoadVal, gpuLoadVal);
+    auto divfOp = loopB.create<arith::DivFOp>(loc, subfOp, cpuLoadVal);
```

Comment:
Hang on a second, I don't think this is the right way to compare floats? Or at least something feels fishy about this bit.

Can we adapt https://stackoverflow.com/questions/4915462/how-should-i-do-floating-point-comparison#answer-32334103 instead?

Reply:
Basically the same except I use golden CPU value instead of adding both. Epsilon varies greatly from f16 to f32.

Ideally I would like the CPU computed with f16 for apples to apples.

Reply:
It looks a lot different, is what I'm noting. I might need to stare at the output to reassure myself, but, well, yeah

And true re eps - I can see that it went down to 1e-1 or 1e-2 and feel comfortable with that - since the epsilons I generally have on my fingers are 1e-5/1e-6 for singles and 1e-9/1e-10 for doubles, so that all progresses nicely by analogy.

Reply:
Ok, yeah, I've stared at it, seems fine.